File: script/make/protocol/cs/ProtocolRouter.cs
public class ProtocolRouter
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Object data)
    {
        switch (protocol / 100)
        {
            case 100: AccountProtocol.Encode(encoding, writer, protocol, data);break;
            case 101: RoleProtocol.Encode(encoding, writer, protocol, data);break;
            case 111: ItemProtocol.Encode(encoding, writer, protocol, data);break;
            case 112: TaskProtocol.Encode(encoding, writer, protocol, data);break;
            case 113: ShopProtocol.Encode(encoding, writer, protocol, data);break;
            case 114: MailProtocol.Encode(encoding, writer, protocol, data);break;
            case 115: FriendProtocol.Encode(encoding, writer, protocol, data);break;
            case 116: ChatProtocol.Encode(encoding, writer, protocol, data);break;
            case 117: SkillProtocol.Encode(encoding, writer, protocol, data);break;
            case 118: BuffProtocol.Encode(encoding, writer, protocol, data);break;
            case 119: TitleProtocol.Encode(encoding, writer, protocol, data);break;
            case 120: FashionProtocol.Encode(encoding, writer, protocol, data);break;
            case 121: BubbleProtocol.Encode(encoding, writer, protocol, data);break;
            case 122: AchievementProtocol.Encode(encoding, writer, protocol, data);break;
            case 123: DailyProtocol.Encode(encoding, writer, protocol, data);break;
            case 150: WelfareProtocol.Encode(encoding, writer, protocol, data);break;
            case 161: AuctionProtocol.Encode(encoding, writer, protocol, data);break;
            case 170: DungeonProtocol.Encode(encoding, writer, protocol, data);break;
            case 180: WarProtocol.Encode(encoding, writer, protocol, data);break;
            case 190: RankProtocol.Encode(encoding, writer, protocol, data);break;
            case 191: RankCenterProtocol.Encode(encoding, writer, protocol, data);break;
            case 192: RankWorldProtocol.Encode(encoding, writer, protocol, data);break;
            case 200: MapProtocol.Encode(encoding, writer, protocol, data);break;
            case 301: GuildProtocol.Encode(encoding, writer, protocol, data);break;
            case 500: NoticeProtocol.Encode(encoding, writer, protocol, data);break;
            case 600: CheatProtocol.Encode(encoding, writer, protocol, data);break;
            case 655: TestProtocol.Encode(encoding, writer, protocol, data);break;
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Object Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol)
    {
        switch (protocol / 100)
        {
            case 100: return AccountProtocol.Decode(encoding, reader, protocol);
            case 101: return RoleProtocol.Decode(encoding, reader, protocol);
            case 111: return ItemProtocol.Decode(encoding, reader, protocol);
            case 112: return TaskProtocol.Decode(encoding, reader, protocol);
            case 113: return ShopProtocol.Decode(encoding, reader, protocol);
            case 114: return MailProtocol.Decode(encoding, reader, protocol);
            case 115: return FriendProtocol.Decode(encoding, reader, protocol);
            case 116: return ChatProtocol.Decode(encoding, reader, protocol);
            case 117: return SkillProtocol.Decode(encoding, reader, protocol);
            case 118: return BuffProtocol.Decode(encoding, reader, protocol);
            case 119: return TitleProtocol.Decode(encoding, reader, protocol);
            case 120: return FashionProtocol.Decode(encoding, reader, protocol);
            case 121: return BubbleProtocol.Decode(encoding, reader, protocol);
            case 122: return AchievementProtocol.Decode(encoding, reader, protocol);
            case 123: return DailyProtocol.Decode(encoding, reader, protocol);
            case 150: return WelfareProtocol.Decode(encoding, reader, protocol);
            case 161: return AuctionProtocol.Decode(encoding, reader, protocol);
            case 170: return DungeonProtocol.Decode(encoding, reader, protocol);
            case 180: return WarProtocol.Decode(encoding, reader, protocol);
            case 190: return RankProtocol.Decode(encoding, reader, protocol);
            case 191: return RankCenterProtocol.Decode(encoding, reader, protocol);
            case 192: return RankWorldProtocol.Decode(encoding, reader, protocol);
            case 200: return MapProtocol.Decode(encoding, reader, protocol);
            case 301: return GuildProtocol.Decode(encoding, reader, protocol);
            case 500: return NoticeProtocol.Decode(encoding, reader, protocol);
            case 600: return CheatProtocol.Decode(encoding, reader, protocol);
            case 655: return TestProtocol.Decode(encoding, reader, protocol);
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}

public class Empty {}

File: script/make/protocol/js/ProtocolRouter.js
import AccountProtocol from "./AccountProtocol.js";
import RoleProtocol from "./RoleProtocol.js";
import ItemProtocol from "./ItemProtocol.js";
import TaskProtocol from "./TaskProtocol.js";
import ShopProtocol from "./ShopProtocol.js";
import MailProtocol from "./MailProtocol.js";
import FriendProtocol from "./FriendProtocol.js";
import ChatProtocol from "./ChatProtocol.js";
import SkillProtocol from "./SkillProtocol.js";
import BuffProtocol from "./BuffProtocol.js";
import TitleProtocol from "./TitleProtocol.js";
import FashionProtocol from "./FashionProtocol.js";
import BubbleProtocol from "./BubbleProtocol.js";
import AchievementProtocol from "./AchievementProtocol.js";
import DailyProtocol from "./DailyProtocol.js";
import WelfareProtocol from "./WelfareProtocol.js";
import AuctionProtocol from "./AuctionProtocol.js";
import DungeonProtocol from "./DungeonProtocol.js";
import WarProtocol from "./WarProtocol.js";
import RankProtocol from "./RankProtocol.js";
import RankCenterProtocol from "./RankCenterProtocol.js";
import RankWorldProtocol from "./RankWorldProtocol.js";
import MapProtocol from "./MapProtocol.js";
import GuildProtocol from "./GuildProtocol.js";
import NoticeProtocol from "./NoticeProtocol.js";
import CheatProtocol from "./CheatProtocol.js";
import TestProtocol from "./TestProtocol.js";

export default class ProtocolRouter {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (Math.trunc(protocol / 100)) {
            case 100: return AccountProtocol.encode(textEncoder, view, offset, protocol, data);
            case 101: return RoleProtocol.encode(textEncoder, view, offset, protocol, data);
            case 111: return ItemProtocol.encode(textEncoder, view, offset, protocol, data);
            case 112: return TaskProtocol.encode(textEncoder, view, offset, protocol, data);
            case 113: return ShopProtocol.encode(textEncoder, view, offset, protocol, data);
            case 114: return MailProtocol.encode(textEncoder, view, offset, protocol, data);
            case 115: return FriendProtocol.encode(textEncoder, view, offset, protocol, data);
            case 116: return ChatProtocol.encode(textEncoder, view, offset, protocol, data);
            case 117: return SkillProtocol.encode(textEncoder, view, offset, protocol, data);
            case 118: return BuffProtocol.encode(textEncoder, view, offset, protocol, data);
            case 119: return TitleProtocol.encode(textEncoder, view, offset, protocol, data);
            case 120: return FashionProtocol.encode(textEncoder, view, offset, protocol, data);
            case 121: return BubbleProtocol.encode(textEncoder, view, offset, protocol, data);
            case 122: return AchievementProtocol.encode(textEncoder, view, offset, protocol, data);
            case 123: return DailyProtocol.encode(textEncoder, view, offset, protocol, data);
            case 150: return WelfareProtocol.encode(textEncoder, view, offset, protocol, data);
            case 161: return AuctionProtocol.encode(textEncoder, view, offset, protocol, data);
            case 170: return DungeonProtocol.encode(textEncoder, view, offset, protocol, data);
            case 180: return WarProtocol.encode(textEncoder, view, offset, protocol, data);
            case 190: return RankProtocol.encode(textEncoder, view, offset, protocol, data);
            case 191: return RankCenterProtocol.encode(textEncoder, view, offset, protocol, data);
            case 192: return RankWorldProtocol.encode(textEncoder, view, offset, protocol, data);
            case 200: return MapProtocol.encode(textEncoder, view, offset, protocol, data);
            case 301: return GuildProtocol.encode(textEncoder, view, offset, protocol, data);
            case 500: return NoticeProtocol.encode(textEncoder, view, offset, protocol, data);
            case 600: return CheatProtocol.encode(textEncoder, view, offset, protocol, data);
            case 655: return TestProtocol.encode(textEncoder, view, offset, protocol, data);
            default:throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (Math.trunc(protocol / 100)) {
            case 100: return AccountProtocol.decode(textDecoder, view, offset, protocol);
            case 101: return RoleProtocol.decode(textDecoder, view, offset, protocol);
            case 111: return ItemProtocol.decode(textDecoder, view, offset, protocol);
            case 112: return TaskProtocol.decode(textDecoder, view, offset, protocol);
            case 113: return ShopProtocol.decode(textDecoder, view, offset, protocol);
            case 114: return MailProtocol.decode(textDecoder, view, offset, protocol);
            case 115: return FriendProtocol.decode(textDecoder, view, offset, protocol);
            case 116: return ChatProtocol.decode(textDecoder, view, offset, protocol);
            case 117: return SkillProtocol.decode(textDecoder, view, offset, protocol);
            case 118: return BuffProtocol.decode(textDecoder, view, offset, protocol);
            case 119: return TitleProtocol.decode(textDecoder, view, offset, protocol);
            case 120: return FashionProtocol.decode(textDecoder, view, offset, protocol);
            case 121: return BubbleProtocol.decode(textDecoder, view, offset, protocol);
            case 122: return AchievementProtocol.decode(textDecoder, view, offset, protocol);
            case 123: return DailyProtocol.decode(textDecoder, view, offset, protocol);
            case 150: return WelfareProtocol.decode(textDecoder, view, offset, protocol);
            case 161: return AuctionProtocol.decode(textDecoder, view, offset, protocol);
            case 170: return DungeonProtocol.decode(textDecoder, view, offset, protocol);
            case 180: return WarProtocol.decode(textDecoder, view, offset, protocol);
            case 190: return RankProtocol.decode(textDecoder, view, offset, protocol);
            case 191: return RankCenterProtocol.decode(textDecoder, view, offset, protocol);
            case 192: return RankWorldProtocol.decode(textDecoder, view, offset, protocol);
            case 200: return MapProtocol.decode(textDecoder, view, offset, protocol);
            case 301: return GuildProtocol.decode(textDecoder, view, offset, protocol);
            case 500: return NoticeProtocol.decode(textDecoder, view, offset, protocol);
            case 600: return CheatProtocol.decode(textDecoder, view, offset, protocol);
            case 655: return TestProtocol.decode(textDecoder, view, offset, protocol);
            default:throw("unknown protocol define: " + protocol)
        }
    }
}